compiler/rustc_target/src/asm/xtensa.rs

```diff
@@ -0,0 +1,244 @@
+        eps6: reg = ["eps6"] % has_exception,
+        eps7: reg = ["eps7"] % has_exception,
+        excsave1: reg = ["excsave1"] % has_exception,
+        excsave2: reg = ["excsave2"] % has_exception,
+        excsave3: reg = ["excsave3"] % has_exception,
+        excsave4: reg = ["excsave4"] % has_exception,
+        excsave5: reg = ["excsave5"] % has_exception,
+        excsave6: reg = ["excsave6"] % has_exception,
+        excsave7: reg = ["excsave7"] % has_exception,
+        exccause: reg = ["exccause"] % has_exception,
+        excvaddr: reg = ["excvaddr"] % has_exception,
+        cpenable: reg = ["cpenable"] % has_coprocessor,
+        vecbase: reg = ["vecbase"] % has_rvector,
+        interrupt: reg = ["interrupt"] % has_interrupt,
+        intclear: reg = ["intclear"] % has_interrupt,
+        intenable: reg = ["intenable"] % has_interrupt,
+        prid: reg = ["prid"] % has_prid,
+        ccount: reg = ["ccount"] % has_timerint,
+        ccompare0: reg = ["ccompare0"] % has_timerint,
+        ccompare1: reg = ["ccompare1"] % has_timerint,
+        ccompare2: reg = ["ccompare2"] % has_timerint,
+        misc0: reg = ["misc0"] % has_miscsr,
+        misc1: reg = ["misc1"] % has_miscsr,
+        misc2: reg = ["misc2"] % has_miscsr,
+        misc3: reg = ["misc3"] % has_miscsr,
+        gpio_out: reg = ["gpio_out"] % has_gpio_out,
+        expstate: reg = ["expstate"] % has_expstate,
+        threadptr: reg = ["threadptr"] % has_threadptr,
+        fcr: reg = ["fcr"] % has_dfpaccel,
+        fsr: reg = ["fsr"] % has_dfpaccel,
+        f64r_lo: reg = ["f64r_lo"] % has_dfpaccel,
+        f64r_hi: reg = ["f64r_hi"] % has_dfpaccel,
+        f64s: reg = ["f64s"] % has_dfpaccel,
+        f0: freg = ["f0"] % has_fp,
+        f1: freg = ["f1"] % has_fp,
+        f2: freg = ["f2"] % has_fp,
+        f3: freg = ["f3"] % has_fp,
+        f4: freg = ["f4"] % has_fp,
+        f5: freg = ["f5"] % has_fp,
+        f6: freg = ["f6"] % has_fp,
+        f7: freg = ["f7"] % has_fp,
+        f8: freg = ["f8"] % has_fp,
+        f9: freg = ["f9"] % has_fp,
+        f10: freg = ["f10"] % has_fp,
+        f11: freg = ["f11"] % has_fp,
+        f12: freg = ["f12"] % has_fp,
+        f13: freg = ["f13"] % has_fp,
+        f14: freg = ["f14"] % has_fp,
+        f15: freg = ["f15"] % has_fp,
+        br: reg = ["br"] % has_bool,
+        b0: breg = ["b0"] % has_bool,
+        b1: breg = ["b1"] % has_bool,
+        b2: breg = ["b2"] % has_bool,
+        b3: breg = ["b3"] % has_bool,
+        b4: breg = ["b4"] % has_bool,
+        b5: breg = ["b5"] % has_bool,
+        b6: breg = ["b6"] % has_bool,
+        b7: breg = ["b7"] % has_bool,
+        b8: breg = ["b8"] % has_bool,
+        b9: breg = ["b9"] % has_bool,
+        b10: breg = ["b10"] % has_bool,
+        b11: breg = ["b11"] % has_bool,
+        b12: breg = ["b12"] % has_bool,
+        b13: breg = ["b13"] % has_bool,
+        b14: breg = ["b14"] % has_bool,
+        b15: breg = ["b15"] % has_bool,
+    }
```

Comment:
Add in these errors
```Rust
        #error = ["a0"] => "a0 is used internally by LLVM and cannot be used as an operand for inline asm",
        #error = ["sp", "a1"] => "sp is used internally by LLVM and cannot be used as an operand for inline asm",

```

Reply:
How does this work when we actually need to change these registers?

i.e: https://github.com/esp-rs/xtensa-lx/blob/c2074c460e70076bc3f29eacdc7607e5c6cc97c1/src/lib.rs#L32-L39

It seems like in that case we can to skirt around this check by not declaring the `sp` as an input/output?

I see in the llvm source they are reserved, but what does reserved actually mean in this context, as we can quite clearly still use them?

Reply:
Good question considering that both a0 and sp are used in asm!() code. I am looking at this comment https://github.com/rust-lang/rust/pull/84732#pullrequestreview-649943225 trying to understand the requirements around adding an asm! target in an attempt to resolve the Todo.

Reply:
Thanks for the link! My understanding from what I've read is that it's meant to be a deterrence, at some point you will *have* to modify these values in order to do anything moderately complex, like building a scheduler for example. In the case of the `set_stack_pointer` function, marking it as unsafe is enough to show the possibility of UB if not used correctly.

wrt `a7`/`a15`, it seems the macro doesn't support the `%` operator on the `#error` branch, i.e `#error = [ "a7"] => "a7 is used internally by LLVM and cannot be used as an operand for inline asm" % has_windowed` doesn't work.

Reply:
Maybe something like what the arm target does?

```Rust
// This uses the same logic as useR7AsFramePointer in LLVM
fn frame_pointer_is_r7(mut has_feature: impl FnMut(&str) -> bool, target: &Target) -> bool {
    target.is_like_osx || (!target.is_like_windows && has_feature("thumb-mode"))
}

fn frame_pointer_r11(
    _arch: InlineAsmArch,
    has_feature: impl FnMut(&str) -> bool,
    target: &Target,
) -> Result<(), &'static str> {
    if !frame_pointer_is_r7(has_feature, target) {
        Err("the frame pointer (r11) cannot be used as an operand for inline asm")
    } else {
        Ok(())
    }
}
```

then in def_regs!
```Rust
        r11: reg = ["r11", "fp"] % frame_pointer_r11,
```